### PR TITLE
Refactor type parameters, implement argument position impl trait

### PR DESCRIPTION
**Pull request by flodiebold**
I wanted to implement APIT by lowering to type parameters because we need to do that anyway for correctness and don't need Chalk support for it; this grew into some more wide-ranging refactoring of how type parameters are handled :sweat_smile: 

 - use Ty::Bound instead of Ty::Param to represent polymorphism, and explicitly
   count binders. This gets us closer to Chalk's way of doing things, and means
   that we now only use Param as a placeholder for an unknown type, e.g. within
   a generic function. I.e. we're never using Param in a situation where we want
   to substitute it, and the method to do that is gone; `subst` now always works
   on bound variables. (This changes how the types of generic functions print; 
   previously, you'd get something like `fn identity<i32>(T) -> T`, but now we
   display the substituted signature `fn identity<i32>(i32) -> i32`, which I think 
   makes more sense.)
 - once we do this, it's more natural to represent `Param` by a globally unique
   ID; the use of indices was mostly to make substituting easier. This also
   means we fix the bug where `Param` loses its name when going through Chalk.
 - I would actually like to rename `Param` to `Placeholder` to better reflect its use and
   get closer to Chalk, but I'll leave that to a follow-up.
 - introduce a context for type lowering, to allow lowering `impl Trait` to
   different things depending on where we are. And since we have that, we can
   also lower type parameters directly to variables instead of placeholders.
   Also, we'll be able to use this later to collect diagnostics.
 - implement argument position impl trait by lowering it to type parameters.
   I've realized that this is necessary to correctly implement it; e.g. consider
   `fn foo(impl Display) -> impl Something`. It's observable that the return
   type of e.g. `foo(1u32)` unifies with itself, but doesn't unify with e.g.
   `foo(1i32)`; so the return type needs to be parameterized by the argument
   type.

   
This fixes a few bugs as well:
 - type parameters 'losing' their name when they go through Chalk, as mentioned
   above (i.e. getting `[missing name]` somewhere)
 - impl trait not being considered as implementing the super traits (very
   noticeable for the `db` in RA)
 - the fact that argument impl trait was only turned into variables when the
   function got called caused type mismatches when the function was used as a
   value (fixes a few type mismatches in RA)

The one thing I'm not so happy with here is how we're lowering `impl Trait` types to variables; since `TypeRef`s don't have an identity currently, we just count how many of them we have seen while going through the function signature. That's quite fragile though, since we have to do it while desugaring generics and while lowering the type signature, and in the exact same order in both cases. We could consider either giving only `TypeRef::ImplTrait` a local id, or maybe just giving all `TypeRef`s an identity after all (we talked about this before)...

Follow-up tasks:
 - handle return position impl trait; we basically need to create a variable and some trait obligations for that variable
 - rename `Param` to `Placeholder`

### CODE REVIEW

**Comment by flodiebold** [comment]
Before:
```
Expressions of unknown type: 5771 (4%)
Expressions of partially unknown type: 3851 (2%)
Type mismatches: 287
```
After:
```
Expressions of unknown type: 5241 (4%)
Expressions of partially unknown type: 3769 (2%)
Type mismatches: 277
```
:tada:

File: crates/ra_hir_ty/src/db.rs [hunk 1]
```diff
@@ -77,6 +77,8 @@ pub trait HirDatabase: DefDatabase {
     #[salsa::interned]
     fn intern_type_ctor(&self, type_ctor: TypeCtor) -> crate::TypeCtorId;
     #[salsa::interned]
+    fn intern_type_param_id(&self, param_id: TypeParamId) -> GlobalTypeParamId;
```
Comment (matklad) [inline crates/ra_hir_ty/src/db.rs]: Why exactly do we need this? `TypeParamId` is also just an ID (although a bigger one), so interning doesn't make a big difference I think?

File: crates/ra_hir_ty/src/lower.rs [hunk 2]
```diff
@@ -101,17 +131,25 @@ impl Ty {
...
                             .collect();
                         Ty::Opaque(predicates)
-                    },
-                    ImplTraitLoweringMode::Placeholder => {
-                        todo!()
-                    },
+                    }
+                    ImplTraitLoweringMode::Variable => {
+                        let idx = ctx.impl_trait_counter.get();
+                        ctx.impl_trait_counter.set(idx + 1);
```
Comment (matklad) [inline crates/ra_hir_ty/src/lower.rs]: These two lines might be extracted into an atomic `alloc_imp_trait_param`

File: crates/ra_hir_ty/src/utils.rs [hunk 3]
```diff
@@ -117,19 +117,31 @@ impl Generics {
...
+
     /// (total, parents, child)
     pub(crate) fn len_split(&self) -> (usize, usize, usize) {
         let parent = self.parent_generics.as_ref().map_or(0, |p| p.len());
         let child = self.params.types.len();
         (parent + child, parent, child)
     }
+
+    /// (self, type param list, impl trait)
+    pub(crate) fn provenance_split(&self) -> (usize, usize, usize) {
```
Comment (matklad) [inline crates/ra_hir_ty/src/utils.rs]: Maybe this should be `-> GenericProvenance`?

**Comment by matklad** [comment]
I'd like to mere this ASAP, seems like it will conflict with anything touching the hir_ty :)

bors r+

File: crates/ra_hir_ty/src/db.rs (hunk 1 above)
Comment (flodiebold) [reply crates/ra_hir_ty/src/db.rs]: TypeParamId are two IDs which we have to fit into one number for Chalk :confused:

**Pull request merged**